### PR TITLE
variational autoencoder

### PR DESCRIPTION
**Pull request by dfalbel**
setting up variational autoencoder
not ready to merge

### CODE REVIEW

**Comment by dfalbel** [comment]
@jjallaire  this is now ready to merge

File: vignettes/examples/variational_autoencoder.R [hunk 1]
```diff
@@ -1 +1,131 @@
...
+    shape = c(batch_size, latent_dim), 
+    mean=0.,
+    stddev=epsilon_std
+  )
+  
+  z_mean + K$exp(z_log_var/2)*epsilon
+}
+
+# note that "output_shape" isn't necessary with the TensorFlow backend
+z <- layer_concatenate(list(z_mean, z_log_var)) %>%
```
Comment (dfalbel) [inline vignettes/examples/variational_autoencoder.R]: @jjallaire here I would like to pass a list of tensors to the lambda layer, but 

`layer_lambda(list(z_mean, z_log_var))` was giving an error `Error: Invalid input to layer function (must be a model or a tensor)`

So I concatenated the values and separated inside the function. Do you see a better approach?

Comment (jjallaire) [inline vignettes/examples/variational_autoencoder.R]: This is the code which receives the layer function input:

https://github.com/rstudio/keras/blob/master/R/layers-core.R#L420-L440

All I can think to do is to check for a list and just call `layer_concatenate` under the hood. However, I'm not sure if that results in an improvement in the code (as I think it's somewhat nice to see the explicit concatenation operation in the code).

Comment (dfalbel) [inline vignettes/examples/variational_autoencoder.R]: How does `layer_concatenate` works? It´s very similar to passing a `list` to `layer_lambda`. The python equivalent code uses a list in this example:

https://github.com/rstudio/keras/blob/master/vignettes/examples/variational_autoencoder.py#L29-L36

Comment (jjallaire) [inline vignettes/examples/variational_autoencoder.R]: After reviewing the equivalent Python code, I think this should be able to work exactly as you are hoping. Could you provide a version of the code which exhibits the `Error: Invalid input` and I'll debug it from there?

Comment (dfalbel) [inline vignettes/examples/variational_autoencoder.R]: Here's a minimal example:

```
x <- layer_input(c(100,2))
y <- layer_input(c(100,2))

my_fun <- function(args){
  a <- args[[1]]
  b <- args[[2]]
  a + b
}

layer_lambda(list(x, y), my_fun)
```

**Comment by jjallaire** [comment]
I think what you want is this then:

```r
library(keras)

x <- layer_input(c(100,2))
y <- layer_input(c(100,2))

my_fun <- function(args){
  a <- args[[1]]
  b <- args[[2]]
  a + b
}

z <- layer_lambda(f = my_fun, arguments = list(x, y))
```

The reason you need the explicit `f` argument is that R keras layers always take a model (or tensor) as their first argument (for magrittr chaining). This isn't required though, but if you want evade this requirement you need to explicitly name your other arguments (that's why in our docs/examples we always name the first argument e.g. `units` in `layer_dense`) so that users become accustomed to always naming layer arguments.

**Comment by dfalbel** [comment]
That's exactly what I needed. I will submit a PR to fix this in the example.

**Pull request merged**